### PR TITLE
Add Typescript Types

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
Have added types to the main file as well as a `tsconfig.json` file. Tests are not working at the moment due to working out mastarm issues.

### CODE REVIEW

**Comment by trevorgerhardt** [comment]
Hi @miles-grant-ibigroup, thanks for contributing. I have a few thoughts/notes on this migration to TypeScript. 

1. `"main": "index.js"`
The main file exported **needs** to be transpiled to JavaScript. 

2. [TSDX](https://tsdx.io/)
I would prefer to adopt TSDX. It's a "zero-config" TypeScript linter, test runner, formatter, and transpiler aimed at packages (as opposed to applications). It should work out of the box for this package and allow us to even remove configuration files keeping everything simple. A new real world example of how it's used can be seen here: https://github.com/placemark/check-geojson/blob/main/package.json

3. [@types/geojson](https://www.npmjs.com/package/@types/geojson)
If possible, let's specifically include `GeoJSON.Point` as an allowable input. It may overlap with how we define `Point` but the type hints around it are useful.

Additionally, we should consider swapping the internal names of `Coordinate` and `Point` to better align with `GeoJSON`.

**Comment by miles-grant-ibigroup** [comment]
Hi Trevor,

Thanks so much for your suggestions! After some discussion we've implemented all of these. 

As to swapping `Coordinate` and `Point`, this would cause breaking changes (`fromCoordinate` and `fromPoint` are backwards, right?) so I think we might hold off on that as part of this migration, unless you have a strong desire not to!

Please let me know if there's anything else you need from me.

File: .github/workflows/node-ci.yml
Comment (evansiroky) [inline .github/workflows/node-ci.yml]: Please also add a step for linting.

Comment (evansiroky) [reply .github/workflows/node-ci.yml]: There also needs to be a step for semantic-release.

File: package.json
Comment (evansiroky) [inline package.json]: Please update to the latest.

File: package.json
Comment (evansiroky) [inline package.json]: This will probably not be needed anymore.

**Review by evansiroky — CHANGES_REQUESTED** [review]
@miles-grant-ibigroup Please add the relevant items to the GitHub Actions script and make the requested package updates as well. This should be good to go pending @trevorgerhardt's approval after that. I'm also tagging @landonreed so he sees what we've done (I can't assign people in this repo).

In our work together, @miles-grant-ibigroup and I think we will probably fork TSDX in order to be able to better implement some custom eslint rules we'd like to use. Surprisingly, TSDX has already effectively stopped being maintained as the original maintainer is now working on a paid version that's supposed to work well with monorepos. 

While it's kind of nice to have jest, lint and transpiling all coming with the same package, any customizations are somewhat difficult. eslint, jest in particular seem to do relatively well on a standalone basis, so perhaps there might be less pain in trying to customize a package that does all 3 of those things than to individually maintain small direct customizations for each within each repo. That also might have an added benefit of not having to change so many things all at once. But I suppose it does run the risk of changes to one of the items breaking the other. Regardless, this TSDX approach here isn't quite perfect, but will probably work well enough for now and we concluded it was probably much less effort than trying to refactor mastarm to do all of this.

**Comment by miles-grant-ibigroup** [comment]
Have made all those fixes thanks! I believe the one last commit to make here is to move to our forked TSDX? Please correct me if I'm wrong.

Finally, I think the CI build is still failing because it's set to use Travis instead of GitHub Actions. I think that will be something @trevorgerhardt has to change.

File: .github/workflows/node-ci.yml
Comment (trevorgerhardt) [inline .github/workflows/node-ci.yml]: Tests should be run on each `push` and `pull_request`, not just on master.

Only semantic release should be restricted to being run on master.

File: tsconfig.json
Comment (trevorgerhardt) [inline tsconfig.json]: Any particular reason we're leaving all the commented out options inline?

**Review by trevorgerhardt — COMMENTED** [review]
Minor nitpicking note on `export` style.

My preference would be:
```js
export function functionName () {}
```
Over
```js
const functionName = () => {}
export { functionName }
```

But I don't think this should hold up the changes. Other comments I have made inline.

File: .gitignore
Comment (evansiroky) [inline .gitignore:41]: We'll need to add a .npmignore file to make sure these type definition files don't get ignored when preparing a release. See [docs](https://docs.npmjs.com/cli/v7/using-npm/developers#keeping-files-out-of-your-package).

File: package.json
Comment (evansiroky) [reply package.json]: Hasn't been updated...

**Review by evansiroky — CHANGES_REQUESTED** [review]
- Please update semantic-release to the latest version (see https://github.com/conveyal/lonlat/pull/19#discussion_r668486515)
- Please create a .npmignore file to make sure type definitions get included.

**Comment by miles-grant-ibigroup** [comment]
Thanks so much for all the polishing suggestions. The weird export style is a relic from when we were figuring out the best way to do the transpiring and still had the cjs exports. Made more sense to have them both there in the same style. But now that those are gone you're right it makes more sense to do it like this!

File: package.json
Comment (miles-grant-ibigroup) [reply package.json:43]: Should've wrote what happened here sorry -- the actual version in the yarn.lock is the newest, but I think since it's still compatible with the version here, the package.json hasn't changed. We could explicitly update the version here if you think that's a good idea?

File: tsconfig.json
Comment (miles-grant-ibigroup) [reply tsconfig.json]: This is leftover from initializing the project as a Typescript project. The entire file can be removed now that things are moved to tsdx!

File: package.json
Comment (miles-grant-ibigroup) [reply package.json:43]: With `semantic-release` it's the same thing. The version is updated in the lock file, but `yarn upgrade` doesn't update the package.json because the latest version is still compatible. Should we explicitly update?

Comment (miles-grant-ibigroup) [reply package.json]: Was doing the upgrades incorrectly, sorry. Should be fixed now!

**Comment by miles-grant-ibigroup** [comment]
Am now working on our tsdx fork. Looks like it has some problems which cause the ci to fail to install packages. The solution is to disable yarn network concurrency, which I've added here for now.

But please let me know if there are any other issues within lonlat to address!

File: package.json
Comment (evansiroky) [inline package.json]: The previous `test` command should still be used here to make sure linting is included.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Please change`test` script and add the `yarn.lock` file to .npmignore. Everything else looks good to me!

Also, we need to make sure there is at least a `fix` or `feat` at the start of a commit message in this PR in order to trigger a new release. So when you make an additional commit be sure to use `fix` or `feat`.

File: .npmignore
Comment (evansiroky) [inline .npmignore]: Please also add `yarn.lock` in here.

**Comment by miles-grant-ibigroup** [comment]
Thanks so much Evan! Have implemented this.

File: .github/workflows/node-ci.yml
Comment (evansiroky) [inline .github/workflows/node-ci.yml]: It's kind of nice to break these out into individual commands so that it is clear which aspect of linting/testing failed in the CI build report.

Comment (miles-grant-ibigroup) [reply .github/workflows/node-ci.yml:24]: Won't it lint twice then?

**Review by evansiroky — APPROVED** [review]
OK, this looks pretty good to me. What do you say @landonreed @binh-dam-ibigroup @trevorgerhardt? Is this mergeable?

**Comment by trevorgerhardt** [comment]
Looks good overall. Any idea why the tests aren't running?

**Comment by miles-grant-ibigroup** [comment]
I think the `conveyal/lonlat` repo is still using travis to run the tests but the travis config was removed and replaced with a GitHub Actions config.x

File: .github/workflows/node-ci.yml
Comment (binh-dam-ibigroup) [inline .github/workflows/node-ci.yml]: This can be removed, and the node version can be inlined in the `with` > `node-version` below, e.g. [see here](https://github.com/opentripplanner/otp-react-redux/blob/dev/.github/workflows/node-ci.yml#L19)

File: index.ts
Comment (binh-dam-ibigroup) [inline index.ts]: This type could be extracted.

File: index.ts
Comment (binh-dam-ibigroup) [inline index.ts]: I personally prefer this syntax for functions (also for all funcs in this file).

```suggestion
export function fromPoint (point: Point): LonLatOutput {
```

**Review by binh-dam-ibigroup — COMMENTED** [review]
I am not really a fan of the long declaration of exported functions `export const toString = function (input: LonLatInput): string {...` but other than that it looks good, see some minor comments.

File: index.ts
Comment (trevorgerhardt) [reply index.ts]: This is my preferred style also.

**Comment by trevorgerhardt** [comment]
> I think the `conveyal/lonlat` repo is still using travis to run the tests but the travis config was removed and replaced with a GitHub Actions config.x

Yes, I guess since it's a fork then Github Actions will not run locally until it's merged.

File: .github/workflows/node-ci.yml
Comment (miles-grant-ibigroup) [reply .github/workflows/node-ci.yml]: Good catch thank you! Fixed in 1711f39

File: index.ts
Comment (miles-grant-ibigroup) [reply index.ts]: All done in 1711f39 thanks!

File: index.ts
Comment (miles-grant-ibigroup) [reply index.ts]: I did this weird double one so that we could have the space between the function name and the parameters (since prettier squishes them otherwise) but I do agree shorter is better. Changed in 1711f39

**Comment by miles-grant-ibigroup** [comment]
Ah that makes sense @trevorgerhardt 
@binh-dam-ibigroup I've made the suggested changes

File: index.ts
Comment (landonreed) [inline index.ts]: Nit: is there a reason this line doesn't use the simpler syntax?
```suggestion
  return { lat, lon }
```

**Review by landonreed — APPROVED** [review]
This looks great. Jumping in at the 11th hour, but everything y'all have done makes sense to me (just one very minor syntax nit).

File: index.ts
Comment (miles-grant-ibigroup) [reply index.ts]: Good catch thank you! I think this may be a relic from before es6.

**Review by trevorgerhardt — APPROVED** [review]
LGTM. Further changes can be made in other PRs.

**Comment by evansiroky** [comment]
> LGTM. Further changes can be made in other PRs.

Thanks @trevorgerhardt! I think you might have to merge this PR since I don't think any of us have write access to this repository.

**Pull request merged**
